components/application/application.js

```diff
@@ -0,0 +1,204 @@
+// libraries
+import { body }  from 'express-validator';
+
+// validator
+import {  protectedRoute } from '../../middleware/protectedRoute.js';
+
+import validationGuard from '../../middleware/validationGuard.js'
+
+// schemas
+import Application from "../../schemas/application.js"
+
+// services
+import "../../services/emailer.js"
+
+
+
+
+async function postMakeApplication(req,res,ext){
+    // if multiple applications aren't allowed, we check if there's a single Application for user id where status is pending
+    // does this need to be a middleware? I don't think so
+
+    try{
+        if(process.env.ALLOW_MULTIPLE_APPLICATIONS !== "true"){
+            let application = await Application.findOne({
+                requestedBy  : req.auth.id,
+                status       : "pending"
+            }).exec()
+    
+            if(application !== null){
+                return res.status(400).json({
+                    err : "User has an application pending review."
+                })
+            }
+        }
+
+        let application = new Application({
+            amount       : parseFloat(req.body.amount),
+            payments     : parseInt  (req.body.payments),
+            paymentAmount: parseFloat(req.body.paymentAmount),
+            reason       : req.body.reason,
+            description  : req.body.description,
+            requestedBy  : req.auth.id,
+            status       : "pending"
+        })
+
+        // client made an invalid request
+        let err = application.validateSync()
+        if(err){
+            return res.status(400).json({
+                err : err.message
+            })
+        }
+
+        await application.save()
+
+        res.status(200).json({
+            msg  : "Application received.",
+        })
+    }catch(e){
+        return next(e)
+    }
+}
+
+
+async function getMyApplications(req,res,next){
+    try{
+        let applications = await Application.find({
+            requestedBy : req.auth.id
+        })
+        // there are some details which we cannot provide to the user
+        applications = applications.map( a => ({
+                // should we wend this one out?
+                id            : a.id,
+                amount        : a.amount,
+                payments      : a.payments,
+                paymentAmount : a.paymentAmount,
+                reason        : a.reason,
+                description   : a.description,
+                status        : a.status
+            })
+        )
+        res.status(200).json(applications)
+    }catch(e){
+        return next(e)
+    }
+}
+
+async function getViewApplication(req,res,next){
+    try{
+        // find application by id and user
+        let a = await Application.findOne({
+            _id         : req.params.id,
+            requestedBy : req.auth.id,

```

Comment:
Req.auth will refer to the authenticated user right? So admins won't be able to use this method since their auth.id will be different?

Reply:
💡We could simply find the applicationById first then:
- If requestor is an admin, return the application
- If requestor is a user, only return the application if it's their own application. Otherwise a 401?

Reply:
Done